File: app/local_plugins/assistant_server/android/src/main/kotlin/com/aiagent/assistant_server/AssistantServerPlugin.kt
package com.aiagent.assistant_server

import android.content.Context
import android.util.Log
import com.aiagent.plugin_interface.NativeAgentConfig
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.*
import java.util.UUID

/**
 * assistant_server Flutter plugin —— Method/Event channel 入口。
 *
 *  - method channel `assistant_server/method`:
 *      • startAssistant(arg: Map) → returns sessionId
 *      • stopActiveSession()
 *      • activeSessionId() → returns String?
 *  - event channel `assistant_server/events`:
 *      消息 / 状态 / 错误 / 连接状态（详见 [AssistantEvents]）
 *
 * 编排逻辑全部在 [AssistantServerCore] / [AssistantSession]，本类只做调度。
 */
class AssistantServerPlugin : FlutterPlugin {

    companion object {
        private const val TAG = "AssistantServerPlugin"
    }

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private val mainScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private lateinit var appContext: Context

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        appContext = binding.applicationContext

        eventChannel = EventChannel(binding.binaryMessenger, "assistant_server/events")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
                AssistantServerCore.bindEmitter { payload ->
                    mainScope.launch { runCatching { events?.success(payload) } }
                }
            }
            override fun onCancel(arguments: Any?) {
                AssistantServerCore.unbindEmitter()
                eventSink = null
            }
        })

        methodChannel = MethodChannel(binding.binaryMessenger, "assistant_server/method")
        methodChannel.setMethodCallHandler { call, result ->
            try {
                handleMethod(call, result)
            } catch (e: Exception) {
                Log.e(TAG, "method ${call.method} failed", e)
                result.error(e::class.java.simpleName, e.message, null)
            }
        }
    }

    private fun handleMethod(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startAssistant" -> {
                val args = call.arguments as? Map<*, *>
                    ?: return result.error("InvalidArgument", "arguments must be a map", null)
                val req = parseRequest(args)
                val sessionId = (args["sessionId"] as? String)
                    ?: ("as_${System.currentTimeMillis()}_${UUID.randomUUID().toString().take(6)}")
                val id = AssistantServerCore.startAssistant(sessionId, req, appContext)
                result.success(id)
            }

            "stopActiveSession" -> {
                AssistantServerCore.stopActive()
                result.success(null)
            }

            "activeSessionId" -> {
                result.success(AssistantServerCore.activeSessionId())
            }

            else -> result.notImplemented()
        }
    }

    private fun parseRequest(args: Map<*, *>): AssistantRequest {
        val agentType = args["agentType"] as? String
            ?: error("agentType required")
        val agentConfigMap = (args["agentConfig"] as? Map<*, *>)
            ?: error("agentConfig required")
        val userLanguage = (args["userLanguage"] as? String) ?: ""

        return AssistantRequest(
            agentType = agentType,
            agentConfig = NativeAgentConfig.fromMap(agentConfigMap),
            userLanguage = userLanguage,
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        AssistantServerCore.unbindEmitter()
        AssistantServerCore.stopActive()
        mainScope.cancel()
    }
}

File: app/local_plugins/assistant_server/ios/Classes/AssistantServerPlugin.swift
import Flutter
import Foundation
import ai_plugin_interface
import os.log

/// assistant_server Flutter plugin（iOS）—— 与 Android `AssistantServerPlugin` 对齐。
///
/// - MethodChannel `assistant_server/method`:
///     • startAssistant(arg: Map) → sessionId
///     • stopActiveSession()
///     • activeSessionId() → String?
/// - EventChannel `assistant_server/events`:
///     消息 / 状态 / 错误 / 连接状态（详见 [AssistantEvents]）
///
/// 编排逻辑在 [AssistantServerCore] / [AssistantSession]，本类只做调度。
public class AssistantServerPlugin: NSObject, FlutterPlugin {

    private static let log = OSLog(subsystem: "com.aiagent.assistant_server", category: "Plugin")

    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private let streamHandler = EventStreamHandler()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = AssistantServerPlugin()
        let messenger = registrar.messenger()

        instance.eventChannel = FlutterEventChannel(
            name: "assistant_server/events",
            binaryMessenger: messenger
        )
        instance.eventChannel?.setStreamHandler(instance.streamHandler)

        instance.methodChannel = FlutterMethodChannel(
            name: "assistant_server/method",
            binaryMessenger: messenger
        )
        registrar.addMethodCallDelegate(instance, channel: instance.methodChannel!)

        // 注入 emitter + device_jieli MethodChannel
        AssistantServerCore.shared.bindContext(messenger: messenger) { payload in
            instance.streamHandler.send(payload)
        }
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        do {
            try dispatch(call, result: result)
        } catch {
            result(FlutterError(code: "AssistantServer", message: error.localizedDescription, details: nil))
        }
    }

    private func dispatch(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        switch call.method {
        case "startAssistant":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(code: "InvalidArgument", message: "arguments must be a map", details: nil))
                return
            }
            let req = try parseRequest(args)
            let sessionId = (args["sessionId"] as? String)
                ?? "as_\(Int64(Date().timeIntervalSince1970 * 1000))_\(UUID().uuidString.prefix(6))"
            do {
                let id = try AssistantServerCore.shared.startAssistant(sessionId: sessionId, request: req)
                result(id)
            } catch let nsErr as NSError {
                result(FlutterError(
                    code: codeFromMessage(nsErr.localizedDescription, defaultCode: "assistant.start_failed"),
                    message: nsErr.localizedDescription,
                    details: nil
                ))
            }

        case "stopActiveSession":
            AssistantServerCore.shared.stopActive()
            result(nil)

        case "activeSessionId":
            result(AssistantServerCore.shared.activeSessionId())

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func parseRequest(_ args: [String: Any]) throws -> AssistantRequest {
        guard let agentType = args["agentType"] as? String, !agentType.isEmpty else {
            throw NSError(domain: "AssistantServer", code: -1,
                          userInfo: [NSLocalizedDescriptionKey: "agentType required"])
        }
        guard let agentConfigMap = args["agentConfig"] as? [String: Any] else {
            throw NSError(domain: "AssistantServer", code: -1,
                          userInfo: [NSLocalizedDescriptionKey: "agentConfig required"])
        }
        let userLanguage = (args["userLanguage"] as? String) ?? ""
        return AssistantRequest(
            agentType: agentType,
            agentConfig: NativeAgentConfig.fromMap(agentConfigMap),
            userLanguage: userLanguage
        )
    }

    private func codeFromMessage(_ msg: String, defaultCode: String) -> String {
        let prefix = msg.split(separator: ":").first.map(String.init) ?? ""
        return prefix.hasPrefix("assistant.") ? prefix : defaultCode
    }
}

/// 把核心层 emit 出来的事件回灌 FlutterEventChannel。
final class EventStreamHandler: NSObject, FlutterStreamHandler {
    private let lock = NSLock()
    private var sink: FlutterEventSink?

    func onListen(withArguments _: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        lock.lock(); sink = events; lock.unlock()
        return nil
    }

    func onCancel(withArguments _: Any?) -> FlutterError? {
        lock.lock(); sink = nil; lock.unlock()
        return nil
    }

    func send(_ payload: [String: Any?]) {
        let s: FlutterEventSink?
        lock.lock(); s = sink; lock.unlock()
        guard let s = s else { return }
        let cleaned = payload.reduce(into: [String: Any]()) { acc, kv in
            if let v = kv.value { acc[kv.key] = v }
        }
        if Thread.isMainThread {
            s(cleaned)
        } else {
            DispatchQueue.main.async { s(cleaned) }
        }
    }
}
